File: Python_Ver/decoding/decodingModel_manual.py
import numpy as np

def decodingModel_manual(M1ensemble):
    """
    Manually designed tuning function to decode movements based on M1 neuron firing rates.

    Parameters:
        M1ensemble : numpy.ndarray
            2D array of shape (4N, T) representing M1 neuron activity, where neurons are grouped in sets of 4.

    Returns:
        decoded_movements : numpy.ndarray
            2D array of shape (3, T) containing:
                Row 1 -> Constant baseline of 0.3
                Row 2 -> Difference between Neuron 1 and Neuron 3 firing rates
                Row 3 -> Difference between Neuron 2 and Neuron 4 firing rates
    """
    # Compute the firing rate of each M1 neuron group (averaging every 4th row)
    M1_fr = np.array([np.mean(M1ensemble[i::4, :], axis=0) for i in range(4)])

    # Construct decoded movements output
    decoded_movements = np.vstack([
        0.3 * np.ones(M1_fr.shape[1]),  # Baseline value
        M1_fr[0, :] - M1_fr[2, :],      # Difference between Neuron 1 and Neuron 3
        M1_fr[1, :] - M1_fr[3, :]       # Difference between Neuron 2 and Neuron 4
    ])

    return decoded_movements

File: decoding/decodingModel_manual.m
function decoded_movements = decodingModel_manual(M1ensemble)
%% Mannual designed tunning function
% get each M1's firing rate
M1_fr = zeros(4, length(M1ensemble));
for i = 1:4
  M1_fr(i, :) = mean(M1ensemble(i:4:end,:));
end

% Rest: Neuron 1 low  & 2 low   & 3 high & 4 high
% Low:  Neuron 1 high & 2 any   & 3 low  & 4 any (Neuron 1 > 3)
% High: Neuron 1 any  & 2 high  & 3 any  & 4 low (Neuron 2 > 4)
decoded_movements = [
  0.3*ones(1,length(M1_fr))
  M1_fr(1,:) - M1_fr(3,:); 
  M1_fr(2,:) - M1_fr(4,:);
];

end
